### ISSUE TITLE
Missing nullability warnings in delegate conversions

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#72253]
**Version Used**:
``` 
d3a41ae7c758b1195f64dc5f57a6eafc8d959469

Merge pull request https://github.com/dotnet/roslyn/pull/72135 from CyrusNajmabadi/delegateCast
Fix remove-unnecessary-cast with necessary delegate cast
```

**Steps to Reproduce**:

Compile and run the following code:
```cs
#nullable enable
using System;

new C().Test(null);

public class C
{
  public void Test(string? maybeNull)
  {
    M(new[] { () => maybeNull });
  }

  private T M<T>(System.Func<T>[] a) where T : notnull
  {
    var result = a[0]();
    Console.WriteLine(result.ToString());
    return result;
  }
}
```

[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4GIYFcA2OLA4CmABETAcQLABQAAgEwCMttMRA7iQMIAUAlADoAKkQDOAF17Y8/ANyt6AZhKMetAN60SqlXQAsJUZN50mABgD8JSAE9gRAHK4c/bSS00dOgLLTOANoAuh4kAiQAvAB8NhD2Ti4kAL7y7kmKOgAOAE5QAG4QEqTCJD4APMJRpkyCdACsFVHBJBD8JBwAFkTZxSQgJDAA9hIyOO6e3iQF2SQ9YrgSkS0B5kECCl7eZgCcvHMLIoMAyhK5MADmAqmbOnQA7LPiCxs66TRJQA)

**Expected Behavior**:
A warning is reported for the method call `M(new[] { () => maybeNull});`

**Actual Behavior**:
No nullability warnings at all are reported in the code above. The code crashes at runtime with a `NullReferenceException`

**Comment by TessenR** [comment]
Ths seems to be just another case of a more general https://github.com/dotnet/roslyn/issues/72274

**Comment by jaredpar** [comment]
Yeah looks like a more specific case there. Closing and will just consider this as part of #72274
